### ISSUE TITLE
Warn if  caller pass json string to PostJsonAsync

### ISSUE THREAD

**Issue by MNF** [tmenier/Flurl#312]
I had a string, that had json presentation of request and I passed it as the parameter of PostJsonAsync. It caused json serialisation of a json string, and server obviously returned bad request. However it took me some time to understand what was wrong, because visually request was similar to the expected. 
It is my mistake, but can you do something to avoid similar confusion in a future?
Could you consider to add some notes in documentation to clarify that the data parameter normally should not be a single string containing json?
Ideally you can generate a warning, if data parameter is a  string containing json.

**Comment by MNF** [comment]
No, exception will be too restrictive. Probably, Debug.WriteLine will be visible for developers and not too annoying, if the passing serialised json string is intentional.

**Comment by tmenier** [comment]
I don't think the documentation is inadequate. Like most things it's documented by examples, and `PostJsonAsync` is demonstrated [here](http://tmenier.github.io/Flurl/), [here](http://tmenier.github.io/Flurl/fluent-http/), and [here](https://github.com/tmenier/Flurl) with an anonymous object.

I don't think  I've ever seen a library use the techniques you're describing to warn that you might be doing something wrong, but if you know of any examples please provide them - I could certainly be wrong about that. (I've learned a ton of things I didn't know from people submitting issues.)

The comments that come through in Intellisense could definitely be better. I'll look at improving those.

**Comment by MNF** [comment]
Examples show one possible way, but they do not warn, that use PostJsonAsync with json string parameter will cause (most likely) unexpected behaviour.

I agree that intellisense notes will be good.

Example that came to my mind is EF Core warning re unneeded includes  (e.g. shown in  https://github.com/aspnet/EntityFrameworkCore/issues/7162 ), which appears in VS Debug Output.
EntityQueryModelVisitor(https://github.com/aspnet/EntityFrameworkCore/blob/0d5f50dc33c9f6cbf3f3567331276022a04d9375/src/EFCore/Query/EntityQueryModelVisitor.cs#L351)  calls
includeCompiler.LogIgnoredIncludes();
The implementation is in LogIgnoredIncludes()
https://github.com/aspnet/EntityFrameworkCore/blob/9c642d3df13a6367d312bbd08a6c440fec012664/src/EFCore/Query/Internal/IncludeCompiler.cs#L125
that calls  IncludeIgnoredWarning 
https://github.com/aspnet/EntityFrameworkCore/blob/dd89d936f2490d4ad3f58f1177ac24195673bfc6/src/EFCore/Extensions/Internal/CoreLoggerExtensions.cs#L405

**Comment by tmenier** [comment]
Better comments are in this release: https://www.nuget.org/packages/Flurl.Http/2.3.2

**Comment by MNF** [comment]
4 years later I hit the same error: I passed a string (json presentation of request) as the parameter of PostJsonAsync instead of using PostStringAsync.
Again it took me some time to understand what was wrong.

**Comment by andrei-micuda** [comment]
I also encountered the same issue. Was under the impression that I could also send the body already serialized to the PostJsonAsync method.
